### ISSUE TITLE
How should intersection typed receivers be handled during extension method resolution?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#3012]
Cf. https://github.com/dart-lang/sdk/issues/52077 for a situation where this question came up. Cf. https://github.com/dart-lang/sdk/issues/52077#issuecomment-1513225793 where @lrhn first clarified the situation.

Consider the following program:

```dart
class Filterable {}

class Filter<T extends Filterable?> {
  void test(T t) {
    if (t is! Filterable) return; // Should promote `t` to `T & Filterable`.
    t.when(); // Accepted by analyzer. Error from CFE.
  }
}

extension When<T extends Filterable> on T {
  T? when() => null;
}

void main() {}
```

The type of `t` at the method invocation `t.when()` is `T & Filterable`. In order to invoke the extension method `when`, the extension needs to receive a type argument, which implies that `T & Filterable` must be erased to a type which is representable at run time.

If it is erased to `T` (which is typical for intersection types when reified), the extension does not match, and we should get a compile time error. This appears to be the approach taken in the CFE.

If it is erased to `Filterable` (possibly after trying `T` and failing), the invocation can resolve to `When.when`, and there are no errors. This may be the approach taken by the analyzer&mdash;at least, it does not report any errors for this program.

Pragmatically, it seems desirable to try each of the two operands of the intersection type. However, it might create difficulties (say, in terms of an exponential time/space usage) because it is a deviation from the one-pass strategies that we are otherwise using.

So what is the desired approach? @dart-lang/language-team, WDYT?

**Comment by lrhn** [comment]
We may not be able to say which erasure is required, but we can require the behavior to be consistent.
Every platform agrees on the constructor invocation, so it's more likely that that's the intended behavior.
(And harder to change).

**Comment by lrhn** [comment]
I agree that the interface argument suggests that the extension members should be based on the promoted type.

The difference is that while a type variable itself has no members, it gets every member except "The Object 5" from its bound or promotion, where the promotion is known to be a subtype of the bound, extension methods can actually add members to _any_ type.
```
extension <T> on T {
  List<T> listOf() => [this]; 
}
void foo<T>(T t1, T t2, void Function(List<T>) callback) {
  if (t1 is int && t1 >= 0) {
    callback(t1.listOf());
  } else {
    callback(t2.listOf());
  }
} 
```
This is a pathological and tortured example, it may very well be that using the promoted bound is always what you want.
It fits shows that it's possible to have extension methods that apply to the base, unbounded and unpromoted type variable type, which won't have any normal interface members.
It's _possible_.

The most _useful_ approach is the selective one: the extension type applies if either part of the intersection type would satisfy the type constraints, then we erase to the part that actually did.
It means trying both. Maybe even introducing a non-trivial rule for which to choose if both match (which could even be based on context type: if the type variable occurs in the context type, definitely choose it).

The reason for all this complexity is that we've taken away your default override strategy. Doing `(x as int).extensionMember` will still keep the receiver type of `x` as `T & int`.
If we take away the syntax you'd use to avoid the problem, we need to avoid it for you.

**Comment by FMorschel** [comment]
I'm not entirely sure if this is a new issue (or is intentional) because it still has intersection types being handled. If it is, please tell me so I can open a new one.

Consider:

```dart
mixin M {
  void m();
}

class M1 with M {
  const M1();
  @override
  void m() {}
}

class A<T> {
  const A(this.o);
  final T o;
}

class B<T extends M> extends A<T> with M {
  const B(super.o);
  @override
  void m() => o.m();
}

void foo<T>(T o) {
  late A<T> a;
  if (o is M) {
    a = B<T>(o); // 'T' doesn't conform to the bound 'M' of the type parameter 'T'. Try using a type that is or is a subclass of 'M'.
  } else {
    a = A<T>(o);
  }
}
```

**Comment by eernstg** [comment]
It is working as intended when we get a compile-time error at `B<T>(o)`: The type test shows that the run-time type of `o` is `M` or a subtype thereof, but `T` can still be anything, for example `Object?`. So we can't conclude anything about the value of `T`.

That's true in general: Promotion says something about the type of a specific object (namely the one which is being tested), and a variable is promotable if we can see at compile time that it will refer to that same object for a while, and then we can give that variable the promoted type. But the type variable has its own value and we can't learn anything about that by testing the type of the object.

**Comment by FMorschel** [comment]
I see, thanks for explaining!

**Comment by FMorschel** [comment]
I'd just don't see why in the case where I don't assign any type to `B` (the second case) it still could not solve. 

I think that's what I'm actually asking here.

If your last proposal (#3831) got through, would that solve this as well?

**Comment by eernstg** [comment]
> Also, if you remove `<T>` there, another error occurs: The argument type 'T & M' can't be assigned to the parameter type 'Never'.

The actual type argument `S` which is chosen by type inference (transforming `B(o)` to `B<S>(o)`) had to have the value `Never`: We must have `S <: T` (such that `B<S> <: A<T>`) and we also need `S <: M` (such that `B<S>` satisfies the declared bound. The greatest lower bound of `T` and `M` is `Never` (that's basically the _only_ lower bound of those two types).

So we're working on `B<Never>(o)`, doing normal type checking, and `o` fails to satisfy the requirement, because the static type of `o` isn't assignable to the type `Never`.

If you're convinced that the surrounding code will behave well (that is, it will satisfy some constraints that aren't expressible in the type system) then you could achieve the same semantics in a way that is guided by some run-time type checks:

```dart
...

void foo<T>(T o) {
  B<S> fooHelper<S extends M>(S o) {
    // In this body the type system knows that `S extends M`,
    // so we can create a `B<S>`.
    return B<S>(o);
  }
  
  late A<T> a;
  if (o is M && <T>[] is List<M>) {
    // At this time we know that `T extends M`. The type system
    // won't recognize that, but we can force it using a dynamic
    // invocation of a generic function.
    a = (fooHelper as Function)<T>(o);
  } else {
    a = A<T>(o);
  }
}

...
```

The crucial part is that we need to know that `T extends M` in order to be allowed to create a `B<T>`, but there is no way we can promote a type variable. So we test this relationship using `<T>[] is List<M>` (which is true if and only if `T extends M`), and then we go ahead and act on it: `fooHelper<T>(o)`. However, we can't use that invocation directly, because the type system doesn't understand that `T extends M`, so we have to do the same thing without compile-time type checking: `(fooHelper as Function)<T>(o)`. This will be checked at run time, but the checks are guaranteed to succeed.

PS: #3831 was a crazy idea, I dropped it again. ;-)

**Comment by eernstg** [comment]
I think that would be interesting! We've never really explored the idea that type parameters could be subject to type tests or type casts in a way which is similar to the instance operations `o is T` and `o as T`, respectively. The most complex part of that effort would probably be to extend the flow analysis such that we don't just get 'true' or 'false' if we evaluate `S <: T` (strawman syntax), but we're also able to use that piece of information to allow expressions that would otherwise have been type errors.

**Comment by eernstg** [comment]
To recapitulate: This is an old issue, and we haven't resolved it. I'd recommend that the interface of the promoted type `B` is  used to determine which members of a receiver of type `X & B` can be invoked, and with which signatures.

**Comment by eernstg** [comment]
@dart-lang/language-team, can we move on this?

**Comment by leafpetersen** [comment]
> @dart-lang/language-team, can we move on this?

I skimmed the issue, I don't see what it is we are proposing to move on.  My understanding is that the implementations behave consistently, and consistently with the model that we've proposed for thinking about this (treat it as constructor inference).  Am I missing something?  Is there an inconsistency we need to resolve?  Otherwise I think we can close this, I'll go ahead and do.  If there's a proposal to make a change, maybe we can pursue it in a clean issue?

**Comment by eernstg** [comment]
> My understanding is that the implementations behave consistently

As I mentioned in the initial message of this issue, we get this response: `// Accepted by analyzer. Error from CFE`. So we do need to make a choice.

> treat it as constructor inference

Ah, of course! It is then an analyzer bug, that is, the analyzer should report an error in the original example. In particular, we do get the error from both tools with this variant:

```dart
class Filterable {}

class Filter<T extends Filterable?> {
  void test(T t) {
    if (t is! Filterable) return; // Should promote `t` to `T & Filterable`.
    When(t).when(); // Rejected by the analyzer and the CFE.
  }
}

class When<T extends Filterable> {
  final T t;
  When(this.t);
  T? when() => null;
}

void main() {}
```

I'll reopen the issue, move it to the SDK, and label it accordingly.
